### ISSUE TITLE
Misclassification on small training set with pre-trained word vectors

### ISSUE THREAD

**Issue by juliohm** [fchollet/keras#3325]
I am one of the many users having issues with passing the pre-trained embedding matrix to the Keras Embedding layer. My code currently looks like the following:

```python
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dropout, Dense
from keras.preprocessing.sequence import pad_sequences
from keras.utils.np_utils import to_categorical

trainX = [sentence_to_vecidxs(entry[0]) for entry in train_data]
trainY = [entry[1] - 1 for entry in train_data]

testX = [sentence_to_vecidxs(entry[0]) for entry in test_data]
testY = [entry[1] - 1 for entry in test_data]

hidden_units = 100
maxwords = 50 # pad sentences to contain this number of words

# padding sentences
trainX = pad_sequences(trainX, maxlen=maxwords)
testX = pad_sequences(testX, maxlen=maxwords)

# class hot vectors
trainY = to_categorical(trainY, nb_classes=nlabels)
testY = to_categorical(testY, nb_classes=nlabels)

# pre-trained word embeddings of size (vocabulary_size x embedding_size)
W = final_embeedings

model = Sequential()
model.add(Embedding(vocabulary_size, embedding_size, input_length=maxwords, mask_zero=False, weights=[W]))
model.add(LSTM(hidden_units, return_sequences=False))
model.add(Dropout(0.3))
model.add(Dense(nlabels, activation='softmax'))

model.compile(loss='categorical_crossentropy', optimizer='sgd', metrics=['accuracy'])
model.fit(trainX, trainY, validation_data=(testX, testY), nb_epoch=30, batch_size=16, verbose=2)
```

The first row of the pre-trained matrix `final_embeddings` contains a random vector for the special token 'PAD'. This token is assigned the index 0 in the corpus and I assume that this index has special meaning in the lookup. The second row in this matrix contains a trained vector for the 'UNK' token which is returned whenever a word is not found in the vocabulary ('PAD' != 'UNK').

In the Embedding layer the input_dim is `vocabulary_size`, which is the same number of rows in the `final_embeddings` matrix. The output_dim is `embedding_size`, which is the dimension of the word vectors. The input_length is `maxwords` which is the (fixed) number of words in the padded sentences (padding with 0). The mask_zero argument is set to False because the matrix `final_embeddings` already has a vector assigned in the first row for the 'PAD' token (index 0).

After running this on a very small training set of ~200 labeled sentences (3 labels) the accuracy is around 55% in both training and validation, but the prediction completely fails even on the training set. Almost all sentences therein are labeled the same.

So now it comes the questions:

1. Is it reasonable to use such a small training set to predict labels on sentences given a pre-trained word embeddnig matrix?
2. We have tried **not** passing the pre-trained embedding matrix and it produces the "same" results in terms of accuracy and predicted labels. How can we make sure this matrix is being used by the LSTM classifier?
3. Is it necessary to have a zero vector for the 'PAD' token or it can be random like we did?
4. We observe that the accuracy is highly correlated to the proportion of the labels in the training set. Do you think this information can help further debug the issue?
5. Could you add a complete example to the documentation with multiclass classification of sentences given word vectors? It seems that many people are struggling to get it working.

Please let me know if you need anything else in order to help with this issue.

**Comment by dolaameng** [comment]
Hi @juliohm, please find some of my thoughts on this:

- Is it reasonable to use such a small training set to predict labels on sentences given a pre-trained word embeddnig matrix?
_Actually I expect pre-trained matrix will specially help when you have a smaller data set, as long as the language features in your classification problem are not so different from those used to train the embedding vectors._

- We have tried not passing the pre-trained embedding matrix and it produces the "same" results in terms of accuracy and predicted labels. How can we make sure this matrix is being used by the LSTM classifier?
_Do you mean that you used an embedding layer without pre-trained matrix or without an embedding layer at all (e.g., by one-hot encoding)?_

- Is it necessary to have a zero vector for the 'PAD' token or it can be random like we did?
_To be frank I haven't totally figured out how this masking works. But usually using a zero vector for padding is a good idea because it essentially correspond to dropping out that part of sequence when you are using LSTM or 1d CNN._

- We observe that the accuracy is highly correlated to the proportion of the labels in the training set. Do you think this information can help further debug the issue?
 _There are many examples within Keras showing how to do multi classification with texts by using different models, e.g., with or without an embeddding, with an LSTM or 1D CNN and etc._

- Could you add a complete example to the documentation with multiclass classification of sentences given word vectors? It seems that many people are struggling to get it working.
_I think the [pretrained_word_embedding example] (https://github.com/fchollet/keras/blob/master/examples/pretrained_word_embeddings.py) is what you are looking for._

_And finally you probably need to set your embedding layer's `trainable=False` as shown in the example above to fix the pretrained embedding vectors in classification. That's what people usually do. And just make sure your sequences in train_X have word hashing index starting from 1 to voacb_size, with 0s as paddings._

**Comment by juliohm** [comment]
Hi @dolaameng, thank you for taking the time. Continuing the discussion:

1. I specifically asked about a training set with only ~200 labeled sentences. Is there any use case with such small number?
2. I meant what I wrote: we kept the embedding layer, but didn't pass the pre-trained matrix.
3. Apparently we have the same question. We both don't know how the mask works and the underlying conventions implemented in the library.
4. I don't see the relation between what I asked and what you wrote.
5. Why the example you linked doesn't mention LSTM?

I hope that my questions are clearer now, I am still at the same page not knowing what is the cause of the issue.

**Comment by dolaameng** [comment]
Hi @juliohm , 

1. whether 200 is too small or not depends on many other things, e.g., distribution of classes (is one of them dominating? ), avg seq len and etc. To be able to finger out which part of model is bringing down the performance, you need to benchmark it with at least a baseline (not necessarily from Keras), e.g., Naive Bayesian, Bag-of-words + MLP and etc. It's hard to tell if another use case is similar to yours just based on the # of sentences and # of classes. 

2. You can get the output from embedding layer and test if your weights are being used.

3. Sorry I thought you were asking about the reason why people usually use zero vector to represent the padding and how a random vector would change the result. That's the part that I am not fully sure of. For the implementation in Keras, you can always refer to the source code.

4. Simply saying "accuracy is highly correlated to the proportion of the labels in the training set. " won't reveal any further information, e.g., if 90% of your sentences are label 1 of course you can get 90% accuracy by just guessing everything is label 1. I was suggesting to use a baseline model to understand whether the problem is on your usage of embedding or other things. Again many Keras examples serve the educational purpose very well.

5. I thought you were asking about "multiclass classification of sentences given word vectors"?  LSTM is only one way of doing text classification. In fact, it is not even accepted as the best way compared to CNN when it comes to classifying the documents.

If your question is not directly related to identified bugs in Keras, I suggest you might want to post your thread to the user discussion group https://groups.google.com/forum/#!forum/keras-users for more answers. Hope it helps!

**Comment by juliohm** [comment]
Hi @dolaameng, thank you for the helpful comments.

1. I have trained a simple neural network on the same dataset, please find the code below:
```python
# Step 9: Train neural net

from sknn.mlp import Classifier, Layer
from sklearn import cross_validation

def sentence_to_vec(sentence_words):
    widxs = [dictionary.get(word, 1) for word in sentence_words]
    return np.mean(final_embeddings[widxs,:], axis=0)

nn = Classifier(layers=[Layer("Tanh", units=20),
                        Layer("Softmax")],
                learning_rate=0.001,
                n_iter=4000)

allvecs = np.array([sentence_to_vec(entry[0]) for entry in interp])
alllabels = np.array([entry[1] for entry in interp])

scores = cross_validation.cross_val_score(nn, allvecs, alllabels, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))
```
The accuracy I get is 0.63 (+/- 0.10), but more importantly, the prediction on unseen data at least gives different labels (3 possible labels), which makes me conclude that something is wrong with my Keras LSTM implementation.

2. Could you please elaborate on how I could perform this check in the original code?

3. Yes, I will end up having to read the source code, this is kind of sub-optimal. I wish this convention was documented somewhere in Keras.

4. It seems that the issue is in the RNN. According to 1.) a simple neural net on the average of vectors is outperforming the LSTM classification.

5. That is interesting, could you please provide an example in the Keras repository that is known to be superior for sentence classification? What neural network architecture is known to work best?

I will definitely refer to the Google forum if I don't get more insight here, thanks for sharing the resource.

**Comment by juliohm** [comment]
I have good news that I would like to share with other users.

If you are not getting good results, definitely consider trying different optimizers and learning rates. It was just this tiny change that caused the entire net to fail in the sentence classification task.

@dolaameng thank you for the support, I am closing the issue.

**Comment by theworkerbee** [comment]
@juliohm Can you share your improved accuracy numbers? I am a little surprised that you got an acceptable classifier with for a 3-class problem with just 200 examples. 

I am also trying to evaluate a similar model with some pre-trained embeddings, so it would interesting to know what other optimizers and learning rates did you try?
